IQSS/dataverse
# IQSS/4959-Support-By-Category-and-Folder-Grouping

**What this PR does / why we need it**:  This PR is an update to the original work in #5485 from QDR in 2019 that added a 'Group by Folder' option for the main dataset file table and an optional 'Goup By Category' option. These work in consort with the main sort options and do a presort/grouping of the files by Category tag (e.g. Documentation, Data, Code, custom tags) and/or Folder before the alphabetical, by age, etc. sorting is done. 

**Which issue(s) this PR closes**:

Closes #4959 
Closes #6933

**Special notes for your reviewer**:
The original work has been in production in QDR since 2019 but was dropped from consideration by IQSS in anticipation of further analysis of other possible options. Since then JHU and ASU have requested this functionality and JHU has supported updating the original branch via GDCC.

**Suggestions on how to test this**: 

- Set the :CategoryOrder setting to something like "Documentation, Data, Code"
- In a dataset with multiple files, sprinkle some category tags and directory path entries
- Verify that the display order puts Documentation first (all files with that tag sorted according to current sort direction) followed by Data (again sorted in that group), etc. and that all files with a folder path appear after the group with no path.
- Verify that setting :OrderByFolder false turns off the by folder sorting
- Set :AllowUserManagementOfOrder true and 
  - Verify that turning the group by toggles on/off turn the groupings on/off
  - Verify that going to page 2 etc with more files or to/from the tree view doesn't mess with the group by checkbox displays or the grouping that is enabled.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
![image](https://user-images.githubusercontent.com/6731983/204922595-3deea588-afa9-4eb9-87b3-ed07b3effc1e.png)

The image shows the two new Group-By-Folder and Group-By-Tag checkbox selectors and shows some of the effect of the tag sort - the second file with the Data tag would normally be third in alphabetical order but has been pulled up to be before untagged files. (Not shown - a file in the /test sub-folder is now last in the list instead of in the top 3 files due to the group-by-folder setting.)

**Is there a release notes update needed for this change?**: in the PR

**Additional documentation**:
